### DISCUSSION TITLE
Error when creating request if any slave is OFF in Modbus RTU

### DISCUSSION THREAD

**Discussion by Thermelgy-Repo** [eModbus/eModbus#162] [eModbus applications]
Dear team,

Thank you for this wonderful library & your continuous effort to make it as the best. Recently while using this library, found a small issue. I have connected 2 slaves via Modbus RTU and it is successfully sending & receiving the request & data. But when I am turning OFF one slave, the polling & recieving signals of other slave is affected. It is showing "  Error creating request: E8 - Request queue full".  After some time it stops the polling part itself. Even if one slave is not working.

I am adding my firmware part & log file too. Kindly help.

[eModbus_Test2.zip](https://github.com/eModbus/eModbus/files/8185888/eModbus_Test2.zip)
[Debug-Log.txt](https://github.com/eModbus/eModbus/files/8185880/Debug-Log.txt)

**Comment by Miq1** [comment]
I am afraid the issue is intrinsically built into your sketch. 
There is just one queue for requests, and any request you are sending will be put at its end to wait for processing.
If a server is not responding, the processing will be held until the defined timeout period has passed and an E0 TIMEOUT response can be returned.
Any requests coming in in the meantime will have to queue up until processing proceeds.
So if you will issue requests in shorter sequence as the timeout period, the request queue will eventually fill up, thus generating an E8 REQUEST QUEUE FULL error and discarding the triggering requests.
If yo look closely at the later lines in your log file, you wil see further E0 error responses, but also data responses from the other server mixed into the E8 responses, because the requests that made it into the queue will be processed one by one.
The simple solution would be to issue requests not faster than with the timeout period in between. This will guarantee the queue will never fill up.
A more elaborate but elegant solution will count the number of consecutive timeouts for each server and reduce the requests to that server accordingly, unless a successful response has been received.

**Reply by Thermelgy-Repo** [reply]
Thank you so much @Miq1 for the clear & detailed explanation. I clearly understand the issue, I have put the polling interval as 1000 ms & Timeout as 2000 ms. I think If I change the polling interval & Timeout to 2000 ms it can be okkey, What timeout interval will be ideal to RTU?.  Also the timeout will be set through setTimeout() function right?

**Reply by Thermelgy-Repo** [reply]
Hi team, 

Now I set the polling interval to 2500 ms & Timeout as 1800ms, After testing I feel it is little improved but still it is showing **"Error creating request: E8 - Request queue full"** error. I am little confused now, So after hitting the timeout before polling interval why the Queue is still filling up. Is there any flag variable to identify the request is sent on process, So that I can check it while polling?. Adding the Debug log & code, Kindly help.

[Debug-Log2.txt](https://github.com/eModbus/eModbus/files/8190258/Debug-Log2.txt)

[eModbus_Test2.zip](https://github.com/eModbus/eModbus/files/8190257/eModbus_Test2.zip)
.

**Reply by Miq1** [reply]
You seem to be sending requests to **all four** of your servers at polling intervals. Since P1 and P2 are not responding, their timeouts will add up. So you will get a waiting time for timeouts of twice 1800ms, which is 3600ms and longer than your polling interval of 2500ms. You are losing 1100ms for each interval, hence after two cycles you will have spent a complete interval additionally.
